api/hasura/cron/epochs.ts

```diff
@@ -230,13 +254,246 @@ export async function notifyEpochEnd({
   return errors;
 }
 
+export async function endEpoch({ endEpoch: { epochs } }: EpochsToNotify) {
+  const endingPromises = epochs.map(async epoch => {
+    const {
+      epoch_pending_token_gifts: pending_gifts,
+      circle,
+      circle_id,
+    } = epoch;
+    assert(circle, `panic: circle #${circle_id} does not exist`);
+
+    // copy pending_gift_tokens to token_gifts
+    // delete pending_gift_tokens
+    if (pending_gifts.length) {
+      await adminClient.mutate({
+        insert_token_gifts: [
+          { objects: pending_gifts },
+          { __typename: true, affected_rows: true },
+        ],
+        delete_pending_token_gifts: [
+          { where: { epoch_id: { _eq: epoch.id } } },
+          { affected_rows: true },
+        ],
+      });
+    }
+
+    const usersWithStartingGive = [] as Array<string>;
+    // if auto_opt_out is true, set users where `starting_tokens == give_tokens_remaining to non_receiver = true
+    // copy user bios to histories
+    // reset give_tokens_received = 0, give_token_remaining = starting tokens, epoch_first_visit = 1, bio = null
+    const userUpdateMutations = circle.users.reduce((ops, user) => {
+      const userUserHasAllGive =
+        user.give_token_remaining === user.starting_tokens;
+      if (userUserHasAllGive) usersWithStartingGive.push(user.name);
+      const optOutMutation =
+        !user.non_receiver && circle.auto_opt_out && userUserHasAllGive
```

Comment:
this should be `!user.non_giver` instead of `!user.non_receiver`

Reply:
This is actually a small optimization to ensure that users who are already opted out don't create an un-needed event trigger request, so I'll keep it and add the non_giver condition on top of it.